File: Loenn/entities/Monumentflipswitch.lua
local drawableSprite = require("structs.drawable_sprite")
local xnaColors = require("consts.xna_colors")
local utils = require("utils")
local Monumentflipswitch = {}

Monumentflipswitch.name = "EmHelper/Monumentflipswitch"
Monumentflipswitch.depth = 2000

Monumentflipswitch.fieldInformation = {
    color = {
        fieldType = "color",
        allowXNAColors = true,
    }
}

local switches = {
    "objects/monumentflipswitch0%s",
    "objects/monumentflipswitch1%s",
    "objects/monumentflipswitch2%s",
    "objects/monumentflipswitch3%s"
}

Monumentflipswitch.placements = {
    {
        name = "Monument Flip Switch (both) [EmHelper]",
        data = {
            onlyEnable = false,
            onlyDisable = false,
            color = "LightSkyBlue",
            mute = false,
            pattern = 0
        },
    },
    {
        name = "Monument Flip Switch (disable) [EmHelper]",
        data = {
            onlyEnable = false,
            onlyDisable = true,
            color = "LightSkyBlue",
            mute = false,
            pattern = 0
        },
    },
    {
        name = "Monument Flip Switch (enable) [EmHelper]",
        data = {
            onlyEnable = true,
            onlyDisable = false,
            color = "LightSkyBlue",
            mute = false,
            pattern = 0
        },
    }
}

function Monumentflipswitch.sprite(room, entity)
    local x, y = entity.x or 0, entity.y or 0
    local onlyEnable = entity.onlyEnable
    local onlyDisable = entity.onlyDisable
    local pattern = (entity.pattern % 4) or 0
    local flipswitchtexture = string.format(switches[pattern+1], "/switch01")

    if onlyEnable then
        flipswitchtexture = string.format(switches[pattern+1], "/switch13")
    elseif onlyDisable then
        flipswitchtexture = string.format(switches[pattern+1], "/switch15")
    end

    local flipswitchSprite = drawableSprite.fromTexture(flipswitchtexture, entity)

    flipswitchSprite:setJustification(1, 1)
    flipswitchSprite:addPosition(16, 16)

    return flipswitchSprite
end

function Monumentflipswitch.rectangle(room, entity)
    return utils.rectangle(entity.x-8, entity.y-6, 16, 21)
end

return Monumentflipswitch

File: Ahorn/entities/Monumentflipswitch.jl
module EmHelperMonumentflipswitch

using ..Ahorn, Maple

@mapdef Entity "EmHelper/Monumentflipswitch" Monumentflipswitch(x::Integer, y::Integer, onlyEnable::Bool=false, onlyDisable::Bool=false, color::String="82d9ff", mute::Bool=false, pattern::Integer=0)

const placements = Ahorn.PlacementDict(
    "Monument FlipSwitch (EmHelper, only enable)" => Ahorn.EntityPlacement(
        Monumentflipswitch,
        "point",
        Dict{String, Any}(
            "onlyEnable" => true
        )
    ),
    "Monument FlipSwitch (EmHelper, only disable)" => Ahorn.EntityPlacement(
        Monumentflipswitch,
        "point",
        Dict{String, Any}(
            "onlyDisable" => true
        )
    ),
    "Monument FlipSwitch (EmHelper, both)" => Ahorn.EntityPlacement(
        Monumentflipswitch
    ),
)

function switchSprite(entity::Monumentflipswitch)
    onlyDisable = get(entity.data, "onlyDisable", false)
    onlyEnable = get(entity.data, "onlyEnable", false)
    patternNumber = get(entity.data, "pattern", 0)
    pattern = string(patternNumber)
    if onlyDisable
        return "objects/monumentflipswitch"*pattern*"/switch13.png"

    elseif onlyEnable
        return "objects/monumentflipswitch"*pattern*"/switch15.png"

    else
        return "objects/monumentflipswitch"*pattern*"/switch01.png"
    end
end

function Ahorn.selection(entity::Monumentflipswitch)
    x, y = Ahorn.position(entity)
    sprite = switchSprite(entity)

    return Ahorn.getSpriteRectangle(sprite, x, y)
end

function getColor(color)
    if haskey(Ahorn.XNAColors.colors, color)
        return Ahorn.XNAColors.colors[color]

    else
        try
            return ((Ahorn.argb32ToRGBATuple(parse(Int, replace(color, "#" => ""), base=16))[1:3] ./ 255)..., 1.0)

        catch

        end
    end

    return (1.0, 1.0, 1.0, 1.0)
end

function Ahorn.render(ctx::Ahorn.Cairo.CairoContext, entity::Monumentflipswitch, room::Maple.Room)
    sprite = switchSprite(entity)

    entitycolor = String(get(entity.data, "color", "82d9ff"))
    entitytint = getColor(entitycolor)

    Ahorn.drawSprite(ctx, sprite, 0, 0, tint=entitytint)
end

end